File: app/src/main/java/com/khmelenko/lab/travisclient/activity/RepoDetailsActivity.java
package com.khmelenko.lab.travisclient.activity;

import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.view.ViewPager;
import android.support.v7.app.ActionBar;
import android.support.v7.widget.Toolbar;
import android.view.View;
import android.widget.Toast;

import com.khmelenko.lab.travisclient.R;
import com.khmelenko.lab.travisclient.TravisApp;
import com.khmelenko.lab.travisclient.adapter.SmartFragmentStatePagerAdapter;
import com.khmelenko.lab.travisclient.fragment.BranchesFragment;
import com.khmelenko.lab.travisclient.fragment.BuildHistoryFragment;
import com.khmelenko.lab.travisclient.fragment.PullRequestsFragment;
import com.khmelenko.lab.travisclient.mvp.MvpActivity;
import com.khmelenko.lab.travisclient.network.response.Branches;
import com.khmelenko.lab.travisclient.network.response.BuildHistory;
import com.khmelenko.lab.travisclient.network.response.Requests;
import com.khmelenko.lab.travisclient.presenter.RepoDetailsPresenter;
import com.khmelenko.lab.travisclient.view.RepoDetailsView;

import javax.inject.Inject;

import butterknife.ButterKnife;

/**
 * Repository Details Activity
 *
 * @author Dmytro Khmelenko
 */
public final class RepoDetailsActivity extends MvpActivity<RepoDetailsPresenter> implements
        RepoDetailsView,
        BuildHistoryFragment.BuildHistoryListener,
        BranchesFragment.BranchesListener,
        PullRequestsFragment.PullRequestsListener {

    private static final int BUILD_DETAILS_REQUEST_CODE = 0;

    public static final String REPO_SLUG_KEY = "RepoSlug";
    public static final String RELOAD_REQUIRED_KEY = "ReloadRequiredKey";

    @Inject
    RepoDetailsPresenter mPresenter;

    private boolean mReloadRequired;
    private boolean mInitialLoad = true;
    private PagerAdapter mAdapterViewPager;

    /**
     * Custom adapter for view pager
     */
    private class PagerAdapter extends SmartFragmentStatePagerAdapter {
        private static final int ITEMS_COUNT = 3;
        private static final int INDEX_BUILD_HISTORY = 0;
        private static final int INDEX_BRANCHES = 1;
        private static final int INDEX_PULL_REQUESTS = 2;

        public PagerAdapter(FragmentManager fragmentManager) {
            super(fragmentManager);
        }

        @Override
        public int getCount() {
            return ITEMS_COUNT;
        }

        @Override
        public Fragment getItem(int position) {
            switch (position) {
                case INDEX_BUILD_HISTORY:
                    return BuildHistoryFragment.newInstance();
                case INDEX_BRANCHES:
                    return BranchesFragment.newInstance();
                case INDEX_PULL_REQUESTS:
                    return PullRequestsFragment.newInstance();
                default:
                    return null;
            }
        }

        @Override
        public CharSequence getPageTitle(int position) {
            switch (position) {
                case INDEX_BUILD_HISTORY:
                    return getString(R.string.repo_details_tab_build_history);
                case INDEX_BRANCHES:
                    return getString(R.string.repo_details_tab_branches);
                case INDEX_PULL_REQUESTS:
                    return getString(R.string.repo_details_tab_pull_requests);
                default:
                    return null;
            }
        }

    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_repo_details);
        ButterKnife.bind(this);
        TravisApp.instance().activityInjector().inject(this);

        initToolbar();

        // setting view pager
        ViewPager vpPager = (ViewPager) findViewById(R.id.repo_details_view_pager);
        mAdapterViewPager = new PagerAdapter(getSupportFragmentManager());
        vpPager.setAdapter(mAdapterViewPager);
        vpPager.setOffscreenPageLimit(PagerAdapter.ITEMS_COUNT);

        TabLayout tabLayout = (TabLayout) findViewById(R.id.repo_details_view_tabs);
        tabLayout.setupWithViewPager(vpPager);
    }

    @Override
    protected RepoDetailsPresenter getPresenter() {
        return mPresenter;
    }

    @Override
    protected void attachPresenter() {
        getPresenter().attach(this);

        String repoSlug = getIntent().getStringExtra(REPO_SLUG_KEY);
        String projectName = repoSlug.substring(repoSlug.indexOf("/") + 1);
        setTitle(projectName);

        getPresenter().setRepoSlug(repoSlug);

        if (mInitialLoad || mReloadRequired) {
            mInitialLoad = false;
            getPresenter().loadData();
        }
    }

    @Override
    public void showProgress() {
        // do nothing, each fragment of the page adapter is responsible for showing progress
    }

    @Override
    public void hideProgress() {
        // do nothing, each fragment of the page adapter is responsible for hiding progress
    }

    @Override
    public void updateBuildHistory(BuildHistory buildHistory) {
        BuildHistoryFragment fragment =
                (BuildHistoryFragment) mAdapterViewPager.getRegisteredFragment(PagerAdapter.INDEX_BUILD_HISTORY);
        fragment.setBuildHistory(buildHistory);
    }

    @Override
    public void updateBranches(Branches branches) {
        BranchesFragment fragment =
                (BranchesFragment) mAdapterViewPager.getRegisteredFragment(PagerAdapter.INDEX_BRANCHES);
        fragment.setBranches(branches);
    }

    @Override
    public void updatePullRequests(Requests requests) {
        PullRequestsFragment fragment =
                (PullRequestsFragment) mAdapterViewPager.getRegisteredFragment(PagerAdapter.INDEX_PULL_REQUESTS);
        fragment.setPullRequests(requests);
    }

    @Override
    public void showBuildHistoryLoadingError(String message) {
        BuildHistoryFragment fragment =
                (BuildHistoryFragment) mAdapterViewPager.getRegisteredFragment(PagerAdapter.INDEX_BUILD_HISTORY);
        fragment.setBuildHistory(null);

        String msg = getString(R.string.error_failed_loading_build_history, message);
        Toast.makeText(this, msg, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void showBranchesLoadingError(String message) {
        BranchesFragment fragment =
                (BranchesFragment) mAdapterViewPager.getRegisteredFragment(PagerAdapter.INDEX_BRANCHES);
        fragment.setBranches(null);

        String msg = getString(R.string.error_failed_loading_branches, message);
        Toast.makeText(this, msg, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void showPullRequestsLoadingError(String message) {
        PullRequestsFragment fragment =
                (PullRequestsFragment) mAdapterViewPager.getRegisteredFragment(PagerAdapter.INDEX_PULL_REQUESTS);
        fragment.setPullRequests(null);

        String msg = getString(R.string.error_failed_loading_pull_requests, message);
        Toast.makeText(this, msg, Toast.LENGTH_SHORT).show();
    }

    /**
     * Initializes toolbar
     */
    private void initToolbar() {
        final Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        final ActionBar actionBar = getSupportActionBar();

        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setDisplayShowHomeEnabled(true);
            toolbar.setNavigationOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    onBackPressed();
                }
            });
        }
    }

    @Override
    public void onBuildSelected(long buildId) {
        goToBuildDetails(buildId);
    }

    @Override
    public void onReloadBuildHistory() {
        getPresenter().loadBuildsHistory();
    }

    @Override
    public void onBranchSelected(long buildId) {
        goToBuildDetails(buildId);
    }

    @Override
    public void onReloadBranches() {
        getPresenter().loadBranches();
    }

    @Override
    public void onPullRequestSelected(long buildId) {
        goToBuildDetails(buildId);
    }

    @Override
    public void onReloadPullRequests() {
        getPresenter().loadRequests();
    }

    /**
     * Navigates to the build details
     *
     * @param buildId Build ID
     */
    private void goToBuildDetails(long buildId) {
        Intent intent = new Intent(this, BuildDetailsActivity.class);
        intent.putExtra(BuildDetailsActivity.EXTRA_BUILD_ID, buildId);
        intent.putExtra(BuildDetailsActivity.EXTRA_REPO_SLUG, getPresenter().getRepoSlug());
        startActivityForResult(intent, BUILD_DETAILS_REQUEST_CODE);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (resultCode == RESULT_OK) {
            if (requestCode == BUILD_DETAILS_REQUEST_CODE) {
                mReloadRequired |= data.getBooleanExtra(BuildDetailsActivity.BUILD_STATE_CHANGED, false);
            }
        }
    }

    @Override
    public void onBackPressed() {
        Intent intent = new Intent();
        intent.putExtra(RELOAD_REQUIRED_KEY, mReloadRequired);
        setResult(RESULT_OK, intent);

        super.onBackPressed();
    }
}

File: app-v3/src/main/java/com/khmelenko/lab/varis/repodetails/RepoDetailsActivity.kt
package com.khmelenko.lab.varis.repodetails

import android.app.Activity
import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.content.Intent
import android.os.Bundle
import android.support.design.widget.TabLayout
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager
import android.support.v4.view.ViewPager
import android.support.v7.widget.Toolbar
import android.widget.Toast
import com.khmelenko.lab.varis.R
import com.khmelenko.lab.varis.activity.BaseActivity
import com.khmelenko.lab.varis.adapter.SmartFragmentStatePagerAdapter
import com.khmelenko.lab.varis.builddetails.BUILD_STATE_CHANGED
import com.khmelenko.lab.varis.builddetails.BuildDetailsActivity
import com.khmelenko.lab.varis.builddetails.EXTRA_BUILD_ID
import com.khmelenko.lab.varis.builddetails.EXTRA_REPO_SLUG
import com.khmelenko.lab.varis.network.response.Branches
import com.khmelenko.lab.varis.network.response.BuildHistory
import com.khmelenko.lab.varis.network.response.Requests
import dagger.android.AndroidInjection
import javax.inject.Inject

private const val BUILD_DETAILS_REQUEST_CODE = 0

const val REPO_SLUG_KEY = "RepoSlug"
const val RELOAD_REQUIRED_KEY = "ReloadRequiredKey"

private const val ITEMS_COUNT = 3
private const val INDEX_BUILD_HISTORY = 0
private const val INDEX_BRANCHES = 1
private const val INDEX_PULL_REQUESTS = 2

/**
 * Repository Details Activity
 *
 * @author Dmytro Khmelenko
 */
class RepoDetailsActivity : BaseActivity(), BuildHistoryFragment.BuildHistoryListener, BranchesFragment.BranchesListener, PullRequestsFragment.PullRequestsListener {

    @Inject
    lateinit var viewModelFactory: RepoDetailsViewModelFactory

    private lateinit var viewModel: RepoDetailsViewModel

    private var reloadRequired: Boolean = false
    private var initialLoad = true
    private lateinit var adapterViewPager: PagerAdapter

    /**
     * Custom adapter for view pager
     */
    private inner class PagerAdapter internal constructor(fragmentManager: FragmentManager) : SmartFragmentStatePagerAdapter(fragmentManager) {

        override fun getCount(): Int {
            return ITEMS_COUNT
        }

        override fun getItem(position: Int): Fragment? {
            return when (position) {
                INDEX_BUILD_HISTORY -> BuildHistoryFragment.newInstance()
                INDEX_BRANCHES -> BranchesFragment.newInstance()
                INDEX_PULL_REQUESTS -> PullRequestsFragment.newInstance()
                else -> null
            }
        }

        override fun getPageTitle(position: Int): CharSequence? {
            return when (position) {
                INDEX_BUILD_HISTORY -> getString(R.string.repo_details_tab_build_history)
                INDEX_BRANCHES -> getString(R.string.repo_details_tab_branches)
                INDEX_PULL_REQUESTS -> getString(R.string.repo_details_tab_pull_requests)
                else -> null
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        AndroidInjection.inject(this)
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_repo_details)

        viewModel = ViewModelProviders.of(this, viewModelFactory).get(RepoDetailsViewModel::class.java)

        initToolbar()

        // setting view pager
        val vpPager = findViewById<ViewPager>(R.id.repo_details_view_pager)
        adapterViewPager = PagerAdapter(supportFragmentManager)
        vpPager.adapter = adapterViewPager
        vpPager.offscreenPageLimit = ITEMS_COUNT

        val tabLayout = findViewById<TabLayout>(R.id.repo_details_view_tabs)
        tabLayout.setupWithViewPager(vpPager)

        init()
    }

    private fun init() {
        val repoSlug = intent.getStringExtra(REPO_SLUG_KEY)
        val projectName = repoSlug.substring(repoSlug.indexOf("/") + 1)
        title = projectName

        viewModel.repoSlug = repoSlug

        if (initialLoad || reloadRequired) {
            initialLoad = false
            viewModel.loadData()
        }

        viewModel.state().observe(this, Observer {
            when (it) {
                is RepoDetailsState.BuildHistoryLoaded -> updateBuildHistory(it.buildHistory)
                is RepoDetailsState.BuildHistoryError -> showBuildHistoryLoadingError(it.message)
                is RepoDetailsState.BranchesLoaded -> updateBranches(it.branches)
                is RepoDetailsState.BranchesError -> showBranchesLoadingError(it.message)
                is RepoDetailsState.PullRequestsLoaded -> updatePullRequests(it.pullRequests)
                is RepoDetailsState.PullRequestsError -> showPullRequestsLoadingError(it.message)
            }
        })
    }

    private fun updateBuildHistory(buildHistory: BuildHistory) {
        val fragment = adapterViewPager.getRegisteredFragment(INDEX_BUILD_HISTORY) as BuildHistoryFragment
        fragment.setBuildHistory(buildHistory)
    }

    private fun updateBranches(branches: Branches) {
        val fragment = adapterViewPager.getRegisteredFragment(INDEX_BRANCHES) as BranchesFragment
        fragment.setBranches(branches)
    }

    private fun updatePullRequests(requests: Requests) {
        val fragment = adapterViewPager.getRegisteredFragment(INDEX_PULL_REQUESTS) as PullRequestsFragment
        fragment.setPullRequests(requests)
    }

    private fun showBuildHistoryLoadingError(message: String?) {
        val fragment = adapterViewPager.getRegisteredFragment(INDEX_BUILD_HISTORY) as BuildHistoryFragment
        fragment.setBuildHistory(null)

        val msg = getString(R.string.error_failed_loading_build_history, message)
        Toast.makeText(this, msg, Toast.LENGTH_SHORT).show()
    }

    private fun showBranchesLoadingError(message: String?) {
        val fragment = adapterViewPager.getRegisteredFragment(INDEX_BRANCHES) as BranchesFragment
        fragment.setBranches(null)

        val msg = getString(R.string.error_failed_loading_branches, message)
        Toast.makeText(this, msg, Toast.LENGTH_SHORT).show()
    }

    private fun showPullRequestsLoadingError(message: String?) {
        val fragment = adapterViewPager.getRegisteredFragment(INDEX_PULL_REQUESTS) as PullRequestsFragment
        fragment.setPullRequests(null)

        val msg = getString(R.string.error_failed_loading_pull_requests, message)
        Toast.makeText(this, msg, Toast.LENGTH_SHORT).show()
    }

    /**
     * Initializes toolbar
     */
    private fun initToolbar() {
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        val actionBar = supportActionBar

        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true)
            actionBar.setDisplayShowHomeEnabled(true)
            toolbar.setNavigationOnClickListener { onBackPressed() }
        }
    }

    override fun onBuildSelected(buildId: Long) {
        goToBuildDetails(buildId)
    }

    override fun onReloadBuildHistory() {
        viewModel.loadBuildsHistory()
    }

    override fun onBranchSelected(buildId: Long) {
        goToBuildDetails(buildId)
    }

    override fun onReloadBranches() {
        viewModel.loadBranches()
    }

    override fun onPullRequestSelected(buildId: Long) {
        goToBuildDetails(buildId)
    }

    override fun onReloadPullRequests() {
        viewModel.loadRequests()
    }

    /**
     * Navigates to the build details
     *
     * @param buildId Build ID
     */
    private fun goToBuildDetails(buildId: Long) {
        val intent = Intent(this, BuildDetailsActivity::class.java)
        intent.putExtra(EXTRA_BUILD_ID, buildId)
        intent.putExtra(EXTRA_REPO_SLUG, viewModel.repoSlug)
        startActivityForResult(intent, BUILD_DETAILS_REQUEST_CODE)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode == Activity.RESULT_OK) {
            if (requestCode == BUILD_DETAILS_REQUEST_CODE && data != null) {
                reloadRequired = reloadRequired or data.getBooleanExtra(BUILD_STATE_CHANGED, false)
            }
        }
    }

    override fun onBackPressed() {
        val intent = Intent()
        intent.putExtra(RELOAD_REQUIRED_KEY, reloadRequired)
        setResult(Activity.RESULT_OK, intent)

        super.onBackPressed()
    }
}
